File: src/core_new/ast.rs
use crate::token::Token;
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum NodeType {
    CompilationUnit,
    Block,
    Statement,
    Expression,
    Function,
    FunctionCall,
    Arguments,
    IfStatement,
    WhenStatement,
    Pattern,
    ForLoop,
    WhileLoop,
    Lambda,
    Ternary,
    Pipe,
    Identifier,
    Literal,
    Operator,
    BinaryOp,
    UnaryOp,
    List,
    Dict,
    MemberAccess,
}

#[derive(Debug, Clone)]
pub struct AstNode {
    pub node_type: NodeType,
    pub token: Option<Token>,
    pub children: Vec<AstNode>,
}

impl AstNode {
    pub fn new(node_type: NodeType, token: Option<Token>) -> Self {
        Self {
            node_type,
            token,
            children: Vec::new(),
        }
    }

    pub fn with_children(node_type: NodeType, token: Option<Token>, children: Vec<AstNode>) -> Self {
        Self {
            node_type,
            token,
            children,
        }
    }

    pub fn add_child(&mut self, child: AstNode) {
        self.children.push(child);
    }

    pub fn token_lexeme(&self) -> String {
        self.token.as_ref().map(|t| t.lexeme.clone()).unwrap_or_default()
    }
}

impl fmt::Display for AstNode {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.display_recursive(f, 0)
    }
}

impl AstNode {
    fn display_recursive(&self, f: &mut fmt::Formatter, depth: usize) -> fmt::Result {
        let indent = "  ".repeat(depth);

        write!(f, "{}{:?}", indent, self.node_type)?;

        if let Some(ref token) = self.token {
            write!(f, " '{}'", token.lexeme)?;
        }

        writeln!(f)?;

        for child in &self.children {
            child.display_recursive(f, depth + 1)?;
        }

        Ok(())
    }
}

// Precedence levels (higher number = tighter binding)
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Precedence(pub i32);

impl Precedence {
    pub const LOWEST: Precedence = Precedence(0);
    pub const ASSIGNMENT: Precedence = Precedence(1);
    pub const PIPE: Precedence = Precedence(2);
    pub const COMMA: Precedence = Precedence(3);
    pub const TERNARY: Precedence = Precedence(4);
    pub const OR: Precedence = Precedence(5);
    pub const AND: Precedence = Precedence(6);
    pub const BITWISE_AND: Precedence = Precedence(7);
    pub const EQUALITY: Precedence = Precedence(8);
    pub const COMPARISON: Precedence = Precedence(9);
    pub const RANGE: Precedence = Precedence(10);
    pub const TERM: Precedence = Precedence(11);
    pub const FACTOR: Precedence = Precedence(12);
    pub const EXPONENT: Precedence = Precedence(13);
    pub const UNARY: Precedence = Precedence(14);
    pub const APPLICATION: Precedence = Precedence(15);
    pub const CALL: Precedence = Precedence(16);
}

File: src/core/quark/ast/ast.go
package ast

import (
	"fmt"
	"quark/token"
	"strings"
)

type NodeType int

const (
	CompilationUnitNode NodeType = iota
	BlockNode
	StatementNode
	ExpressionNode
	ConditionNode
	FunctionNode
	FunctionCallNode
	ArgumentsNode
	IdentifierNode
	LiteralNode
	OperatorNode
	IfStatementNode
	WhenStatementNode
	PatternNode
	ForLoopNode
	WhileLoopNode
	LambdaNode
	TernaryNode
	PipeNode
	ListNode
	DictNode
	IndexNode
	ModuleNode
	UseNode
)

var nodeTypeNames = map[NodeType]string{
	CompilationUnitNode: "CompilationUnit",
	BlockNode:           "Block",
	StatementNode:       "Statement",
	ExpressionNode:      "Expression",
	ConditionNode:       "Condition",
	FunctionNode:        "Function",
	FunctionCallNode:    "FunctionCall",
	ArgumentsNode:       "Arguments",
	IdentifierNode:      "Identifier",
	LiteralNode:         "Literal",
	OperatorNode:        "Operator",
	IfStatementNode:     "IfStatement",
	WhenStatementNode:   "WhenStatement",
	PatternNode:         "Pattern",
	ForLoopNode:         "ForLoop",
	WhileLoopNode:       "WhileLoop",
	LambdaNode:          "Lambda",
	TernaryNode:         "Ternary",
	PipeNode:            "Pipe",
	ListNode:            "List",
	DictNode:            "Dict",
	IndexNode:           "Index",
	ModuleNode:          "Module",
	UseNode:             "Use",
}

func (n NodeType) String() string {
	if name, ok := nodeTypeNames[n]; ok {
		return name
	}
	return "Unknown"
}

// Node is the interface all AST nodes implement
type Node interface {
	TokenLiteral() string
	String() string
	Type() NodeType
}

// TreeNode is the main AST node structure (mirrors Python implementation)
type TreeNode struct {
	NodeType NodeType
	Token    *token.Token
	Children []*TreeNode
}

func NewNode(nodeType NodeType, tok *token.Token) *TreeNode {
	return &TreeNode{
		NodeType: nodeType,
		Token:    tok,
		Children: make([]*TreeNode, 0),
	}
}

func (n *TreeNode) Type() NodeType {
	return n.NodeType
}

func (n *TreeNode) TokenLiteral() string {
	if n.Token != nil {
		return n.Token.Literal
	}
	return ""
}

func (n *TreeNode) String() string {
	if n.Token != nil && n.Token.Literal != "" {
		return fmt.Sprintf("%s[%s]", n.NodeType.String(), n.Token.Literal)
	}
	return n.NodeType.String()
}

func (n *TreeNode) AddChild(child *TreeNode) {
	n.Children = append(n.Children, child)
}

func (n *TreeNode) AddChildren(children ...*TreeNode) {
	n.Children = append(n.Children, children...)
}

// Print outputs the AST tree structure
func (n *TreeNode) Print(level int) {
	indent := strings.Repeat("  ", level)
	fmt.Printf("%s%s\n", indent, n.String())
	for _, child := range n.Children {
		child.Print(level + 1)
	}
}

// PrintTree is a convenience function for printing from root
func (n *TreeNode) PrintTree() {
	n.Print(0)
}

// Precedence levels for expression parsing (mirrors Python Precedence class)
type Precedence int

const (
	PrecLowest     Precedence = iota
	PrecAssignment            // =
	PrecPipe                  // |
	PrecComma                 // ,
	PrecTernary               // if-else
	PrecOr                    // or
	PrecAnd                   // and
	PrecBitwiseAnd            // &
	PrecEquality              // == !=
	PrecComparison            // < <= > >=
	PrecRange                 // ..
	PrecTerm                  // + -
	PrecFactor                // * / %
	PrecExponent              // **
	PrecUnary                 // ! ~ -
	PrecApplication           // function application (space)
	PrecAccess                // . [] ()
)

var precedenceNames = map[Precedence]string{
	PrecLowest:     "Lowest",
	PrecAssignment: "Assignment",
	PrecPipe:       "Pipe",
	PrecComma:      "Comma",
	PrecTernary:    "Ternary",
	PrecOr:         "Or",
	PrecAnd:        "And",
	PrecBitwiseAnd: "BitwiseAnd",
	PrecEquality:   "Equality",
	PrecComparison: "Comparison",
	PrecRange:      "Range",
	PrecTerm:       "Term",
	PrecFactor:     "Factor",
	PrecExponent:   "Exponent",
	PrecUnary:      "Unary",
	PrecApplication: "Application",
	PrecAccess:     "Access",
}

func (p Precedence) String() string {
	if name, ok := precedenceNames[p]; ok {
		return name
	}
	return "Unknown"
}
